packages/core/src/controllers/ConvertController.ts

```diff
@@ -0,0 +1,687 @@
+  inputError: undefined,
+
+  // Request values
+  slippage: ConstantsUtil.CONVERT_SLIPPAGE_TOLERANCE,
+
+  // Tokens
+  tokens: undefined,
+  popularTokens: undefined,
+  suggestedTokens: undefined,
+  foundTokens: undefined,
+  myTokensWithBalance: undefined,
+  tokensPriceMap: {},
+
+  // Calculations
+  gasFee: BigInt(0),
+  gasPriceInUSD: 0,
+  priceImpact: undefined,
+  maxSlippage: undefined
+})
+
+// -- Controller ---------------------------------------- //
+export const ConvertController = {
+  state,
+
+  subscribe(callback: (newState: ConvertControllerState) => void) {
+    return sub(state, () => callback(state))
+  },
+
+  subscribeKey<K extends StateKey>(key: K, callback: (value: ConvertControllerState[K]) => void) {
+    return subKey(state, key, callback)
+  },
+
+  getParams() {
+    const { address } = AccountController.state
+
+    if (!address) {
+      throw new Error('No address found to swap the tokens from.')
+    }
+
+    return {
+      fromAddress: address as `0x${string}`,
+      sourceTokenAddress: state.sourceToken?.address,
+      toTokenAddress: state.toToken?.address,
+      toTokenAmount: state.toTokenAmount,
+      toTokenDecimals: state.toToken?.decimals,
+      sourceTokenAmount: state.sourceTokenAmount,
+      sourceTokenDecimals: state.sourceToken?.decimals
+    }
+  },
+
+  setLoading(loading: boolean) {
+    state.loading = loading
+  },
+
+  setSourceToken(sourceToken: TokenInfo | undefined) {
+    if (!sourceToken) {
+      return
+    }
+
+    state.sourceToken = sourceToken
+    this.setTokenValues(sourceToken.address, 'sourceToken')
+  },
+
+  setSourceTokenAmount(amount: string) {
+    const { sourceTokenAddress } = this.getParams()
+
+    state.sourceTokenAmount = amount
+
+    if (sourceTokenAddress) {
+      this.setTokenValues(sourceTokenAddress, 'sourceToken')
+    }
+  },
+
+  setToToken(toToken: TokenInfo | undefined) {
+    const { sourceTokenAddress, sourceTokenAmount } = this.getParams()
+
+    if (!toToken) {
+      return
+    }
+
+    state.toToken = toToken
+    this.setTokenValues(toToken.address, 'toToken')
+
+    if (sourceTokenAddress && sourceTokenAmount) {
+      this.makeChecks()
+    }
+  },
+
+  setToTokenAmount(amount: string) {
+    const { toTokenAddress } = this.getParams()
+
+    state.toTokenAmount = amount
+
+    if (toTokenAddress) {
+      this.setTokenValues(toTokenAddress, 'toToken')
+    }
+  },
+
+  async setTokenValues(address: string, target: 'sourceToken' | 'toToken') {
+    let price = parseFloat(state.tokensPriceMap[address] || '0')
+
+    if (!price) {
+      price = await this.getAddressPrice(address)
+    }
+
+    if (target === 'sourceToken') {
+      state.sourceTokenPriceInUSD = price
+    } else if (target === 'toToken') {
+      state.toTokenPriceInUSD = price
+    }
+  },
+
+  switchTokens() {
+    const newSourceToken = state.toToken ? { ...state.toToken } : undefined
+    const newToToken = state.sourceToken ? { ...state.sourceToken } : undefined
```

Comment:
It would work, but with issues. Due to JavaScript's nature, assigning one variable which points to an object to another variable is not the same when assigning variables which references to primitive values like `string`, `number`, and `boolean`.

JavaScript working with pointers and when we assign one object to another, we are actually assigning a reference to the first object. This means both variables point to the same memory location. Consequently, if you modify the object through one variable, the changes will be reflected when accessing it through the other variable. This can lead to unintended side effects, especially in complex applications where it might be hard to track which parts of your code are modifying the object.

Destructuring, on the other hand, is useful because it allows us to create a new object with the same properties but with a separate reference. This avoids the issues associated with direct object references. When you destructure and assign the properties of one object to another new object, you are essentially copying the values. For primitives, this is straightforward, and for objects, if done carefully, it can help ensure that the objects don't share the same reference.